### PR TITLE
Scrapper, Tasenko Olga - 21FPL1

### PR DESCRIPTION
**Pull request by Tasenko03**

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -0,0 +1,2 @@
+requests==2.28.2
```
Comment (demid5111) [inline requirements.txt:3]: where are changes in config.json?

File: lab_5_scrapper/target_score.txt [hunk 2]
```diff
@@ -1 +1 @@
-6
\ No newline at end of file
+4
```
Comment (demid5111) [inline lab_5_scrapper/target_score.txt]: too small

File: lab_5_scrapper/scrapper.py [hunk 3]
```diff
@@ -90,25 +193,36 @@ def __init__(self, config: Config) -> None:
...
+        self._seed_url = config.get_seed_urls()
+        self._config = config
+        self.urls = []
 
     def _extract_url(self, article_bs: BeautifulSoup) -> str:
         """
         Finds and retrieves URL from HTML
         """
-        pass
+        href = article_bs.get('href')
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py]: unnecessary variable

File: lab_5_scrapper/scrapper.py [hunk 4]
```diff
@@ -13,70 +68,118 @@ def __init__(self, path_to_config: Path) -> None:
...
+        return self._headless_mode
 
 
 def make_request(url: str, config: Config) -> requests.models.Response:
     """
     Delivers a response from a request
     with given configuration
     """
-    pass
+    time.sleep(random.randint(TIMEOUT_LOWER_LIMIT, TIMEOUT_UPPER_LIMIT))
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py]: nice! 🏅

File: lab_5_scrapper/scrapper.py [hunk 5]
```diff
@@ -13,70 +68,118 @@ def __init__(self, path_to_config: Path) -> None:
...
+    headers = config.get_headers()
+    timeout = config.get_timeout()
+
+    response = requests.get(url, headers=headers, timeout=timeout)
+    return response
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py:187]: set encoding before return

File: lab_5_scrapper/scrapper.py [hunk 6]
```diff
@@ -13,70 +68,118 @@ def __init__(self, path_to_config: Path) -> None:
...
+        seed_urls = config['seed_urls']
+        headers = config['headers']
+        total_articles_to_find_and_parse = config['total_articles_to_find_and_parse']
+        encoding = config['encoding']
+        timeout = config['timeout']
+        verify_certificate = config['should_verify_certificate']
+        headless_mode = config['headless_mode']
+
+        for url in seed_urls:
+            if not re.match(re.compile(r"https?://w?w?w?."), url):
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py]: do you think it is possible to have for example just one w?

File: lab_5_scrapper/scrapper.py [hunk 7]
```diff
@@ -13,70 +68,118 @@ def __init__(self, path_to_config: Path) -> None:
...
+            config = json.load(f)
+        return ConfigDTO(**config)
 
     def _validate_config_content(self) -> None:
         """
         Ensure configuration parameters
         are not corrupt
         """
-        pass
+        with open(self.path_to_config, 'r', encoding='utf-8') as f:
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py]: duplicated read

File: lab_5_scrapper/scrapper.py [hunk 8]
```diff
@@ -13,70 +68,118 @@ def __init__(self, path_to_config: Path) -> None:
         """
         Initializes an instance of the Config class
         """
-        pass
+        self.path_to_config = path_to_config
+        self._validate_config_content()
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py:76]: in init:

1. you need to extract config's content with a function
2. run validation using DTO from the previuos step
3. fill attributes of config and continue work

File: lab_5_scrapper/scrapper.py [hunk 9]
```diff
@@ -13,70 +68,118 @@ def __init__(self, path_to_config: Path) -> None:
...
     def _validate_config_content(self) -> None:
         """
         Ensure configuration parameters
         are not corrupt
         """
-        pass
+        with open(self.path_to_config, 'r', encoding='utf-8') as f:
+            config = json.load(f)
+
+        seed_urls = config['seed_urls']
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py]: rework validation to operate on top of DTO instance

File: lab_5_scrapper/scrapper.py [hunk 10]
```diff
@@ -151,7 +265,9 @@ def prepare_environment(base_path: Union[Path, str]) -> None:
     """
     Creates ASSETS_PATH folder if no created and removes existing folder
     """
-    pass
+    if base_path.exists():
+        shutil.rmtree(base_path)
+    base_path.mkdir(parents=True)
 
 
 def main() -> None:
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py:333]: add demonstration logic, ensure it will be able to collect 100 articles

**Review by demid5111 — CHANGES_REQUESTED** [review]
hurry up, progress is nice, but time is going

File: lab_5_scrapper/scrapper.py [hunk 11]
```diff
@@ -13,70 +71,117 @@ def __init__(self, path_to_config: Path) -> None:
...
         are not corrupt
         """
-        pass
+        config = self._extract_config_content()
+
+        if not isinstance(config.seed_urls, list):
+            raise IncorrectSeedURLError('seed URL is not a list')
+
+        for url in config.seed_urls:
+            if not re.match(re.compile(r"https?://w?."), url):
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py]: now you accept sites like http://w.google.com is that correct?

File: lab_5_scrapper/scrapper.py (hunk 4 above)
Comment (demid5111) [reply lab_5_scrapper/scrapper.py]: but now I see that it is a bit wrong. timeout like `TIMEOUT_UPPER_LIMIT` should be used in checks, not here. here you need to provide some random sleep in range of 1-3

File: lab_5_scrapper/scrapper.py [hunk 12]
```diff
@@ -90,25 +195,49 @@ def __init__(self, config: Config) -> None:
...
+            return href
+        return "None"
 
     def find_articles(self) -> None:
         """
         Finds articles
         """
-        pass
+        for url in self._seed_url:
+            driver = webdriver.Chrome(executable_path="C:\\Users\\Ольга\\Desktop\\2022-2-level-ctlr\\chrome "
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py]: ???

File: lab_5_scrapper/scrapper.py [hunk 13]
```diff
@@ -90,25 +195,49 @@ def __init__(self, config: Config) -> None:
...
+                                              "driver\\chromedriver_win32\\chromedriver.exe")
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py]: 1. discard dynamic crawling
2. even if you want (and I wholeheartedly convince you to give up) such paths specific to your machine will never work in CI

File: lab_5_scrapper/scrapper.py [hunk 14]
```diff
@@ -120,13 +249,22 @@ def __init__(self, full_url: str, article_id: int, config: Config) -> None:
...
     def _fill_article_with_text(self, article_soup: BeautifulSoup) -> None:
         """
         Finds text of article
         """
-        pass
+        elements = article_soup.find_all("div", class_='mb-[24px] lg:mb-[28px]')
+        paragraphs = ""
+        for elem in elements:
+            par = elem.find_all('p')
+            paragraphs += " ".join([text.get_text(strip=True) for text in par])
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py]: join once, avoid string concatenation

File: lab_5_scrapper/scrapper_config.json [hunk 15]
```diff
@@ -1,9 +1,9 @@
 {
-    "seed_urls": [],
+    "seed_urls": ["https://prmira.ru"],
     "headers": {},
```
Comment (demid5111) [inline lab_5_scrapper/scrapper_config.json]: where are haeaders

**Review by demid5111 — CHANGES_REQUESTED** [review]
some more work is required. keep going

File: lab_5_scrapper/scrapper.py (hunk 5 above)
Comment (Tasenko03) [reply lab_5_scrapper/scrapper.py:187]: Done

File: requirements.txt (hunk 1 above)
Comment (Tasenko03) [reply requirements.txt:3]: Added

File: lab_5_scrapper/scrapper.py (hunk 8 above)
Comment (Tasenko03) [reply lab_5_scrapper/scrapper.py:76]: Done

**Review by demid5111 — CHANGES_REQUESTED** [review]
https://prmira.ru/news/2023-04-13/tsifrovoy-bank-tochka-vystavlen-na-prodazhu-2902423 not all text collected

**Comment by demid5111** [comment]
nothing to review to this moment

**Comment by vitazaytseva** [comment]
Well done! You can proceed to work on Lab 6)

**Pull request closed without merging**